### PR TITLE
Enumerate lints for analysis in synthesized options file.

### PR DESCRIPTION
**Pull request by pq**
This enables us to exclude `library_names` pending a new stable SDK push (see notes: https://github.com/flutter/flutter/pull/1996).  Also excluded are a few lints who's feedback was just getting regexp'ed out.  Moving forward, this will allow us to selectively enable specific lints of interest as well.

### CODE REVIEW

**Comment by eseidelGoogle** [comment]
This is best reviewed by @Hixie or @devoncarew I suspect.  Do we want to be generating this options file, or depending on the flutter project to include one (i.e. changing `flutter create` instead?)

**Comment by devoncarew** [comment]
Having `flutter analyze` use an options file will let it analyze code that doesn't import the `flutter` package. So, `create` shouldn't create an `.analysis_options` file; it should depend on the analyze picking up the right config from the embedder.yaml file.

@pq, lgtm. Generating an analysis options file in place works. We could also check one in (~`packages/flutter_tools/lib/src/commands/.analysis_options`?), and then reference the file in the cli arg relative to the flutter root (`ArtifactStore.flutterRoot`). I don't have a preference for either one Having it be a separate file might make it slightly easier to maintain.

**Comment by pq** [comment]
Adding a file to `packages/flutter_tools/lib/src/commands` works for me too though in some ways it's handy to have all of the configuration in one place (e.g., analyzer command line args, options and regexps). Curious if @Hixie has a strong opinion either way.

File: packages/flutter_tools/lib/src/commands/analyze.dart [hunk 1]
```diff
@@ -252,7 +273,8 @@ class AnalyzeCommand extends FlutterCommand {
       '--fatal-warnings',
       '--strong-hints',
       '--fatal-hints',
-      '--lints',
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/analyze.dart]: doesn't this mean we'll miss new lints as they get added?

**Comment by Hixie** [comment]
I think having the file inline (as in this patch) makes more sense from a maintenance point of view.

However, I don't think we should be doing a whitelist of lints. We should be doing a blacklist. A whitelist means we'll miss new lints as they come around.

**Comment by pq** [comment]
> However, I don't think we should be doing a whitelist of lints. We should be doing a blacklist. A whitelist means we'll miss new lints as they come around.

The thinking has been that rulesets will, in practice, be tailored to projects and teams and at the moment the list of lints defined as a default is very small and slowly changing.  I hope we can grow it significantly soon but where there are matters of style there's always push-back.  (Note also that style aside, picking a ruleset will require some choices since a number of them are mutually incompatible.)  In the short-term I propose we manage flutter inclusions manually.  I don't have a great answer to the discovery problem long-term but am certainly open to suggestions.

**Comment by Hixie** [comment]
Is there a way to do a blacklist?

If there's not, and since we're generating the file anyway, is there a way to obtain the current authoritative  list of lints and just include them all except the ones we're excluding?

There might be push-back in other projects but here we're all about getting the most we can get out of the analyzer and I'd much rather blacklist individual inapplicable lints when we roll dart than have to require that whoever rolls dart also be responsible for figuring out which lints have been added (or indeed, that anyone have to do this).

**Comment by pq** [comment]
> If there's not, and since we're generating the file anyway, is there a way to obtain the current authoritative list of lints and just include them all except the ones we're excluding?

If we're willing to add a dependency to the `linter` package this would be easy to do programmatically.  Let me know if you want to go that path and I'll throw together a follow-up PR.

**Comment by Hixie** [comment]
What I'd prefer is if it was just a blacklist.

How does the linter package relate to the analyzer package? (We already import the analyzer package, via test.)

**Comment by pq** [comment]
The linter is a plugin that gets added by analyzer clients (e.g., the `dartanalyzer` CLI tool and the `analysis_server`). Besides defining a `LintCode` and managing an association of enabled lints to analysis contexts, the `analyzer` package is unaware of contributed lint rules.

**Comment by Hixie** [comment]
Is it part of the Dart SDK then? If so, we already have it, so depending on it doesn't make our life worse (you can even probably use a path dependency to bring it in from the local SDK copy rather than from pub).

But I still think we should go with a blacklist. Does the `.analysis_options` file not support blacklisting? How would you exclude a specific warning or error? Are lints opt-in while errors are opt-out?

**Comment by pq** [comment]
It is part of the SDK.  It's snapshotted so I'd need to add the source to make a path dep work but that's easily done.

> But I still think we should go with a blacklist.

:+1: 

> Does the .analysis_options file not support blacklisting? How would you exclude a specific warning or error?

```
analyzer:
  errors:
    invalid_assignment: ignore
    unused_local_variable: error
```

> Are lints opt-in while errors are opt-out?

Yes.

I've been thinking about this a bit and it may be reasonable to add support to the options file and be done with it.  I'll follow up with a doc to spitball ideas.

In the meantime, how do you feel about landing this change with the promise of follow-up improvements?

**Comment by Hixie** [comment]
Sure. Can you add a comment explaining that lints are opt-in, not opt-out, but that this is as of today the complete list and put a link to a bug that's tracking making this not require regular maintenance?

LGTM.

**Comment by pq** [comment]
Absolutely.  Here's the bug with a link to a doc for gratuitous bike-shedding :smile: : https://github.com/dart-lang/sdk/issues/25843

**Pull request merged**